### PR TITLE
Add android native ble library

### PR DESCRIPTION
**Pull request by mbertin-ledger**

### CODE REVIEW

**Review by juan-cortes — APPROVED** [review]

File: bletransportsample/src/main/java/com/ledger/live/bletransportsample/screen/scan/ScanViewModel.kt
Comment (smartine-ledger) [inline bletransportsample/src/main/java/com/ledger/live/bletransportsample/screen/scan/ScanViewModel.kt]: [SHOULD] when working with coroutines : it's recommended to inject the dispatcher to make UT easy to write . In addition VM methods should be safe to be called from the main thread. This means it should not be responsible to change the dispatcher

File: bletransportsample/src/main/java/com/ledger/live/bletransportsample/screen/scan/ScanViewModel.kt
Comment (smartine-ledger) [inline bletransportsample/src/main/java/com/ledger/live/bletransportsample/screen/scan/ScanViewModel.kt]: [SHOULD]  inject BLEManager and avoid static call to make UT easy to write

File: bletransportsample/src/main/java/com/ledger/live/bletransportsample/screen/scan/ScanViewModel.kt
Comment (smartine-ledger) [inline bletransportsample/src/main/java/com/ledger/live/bletransportsample/screen/scan/ScanViewModel.kt]: [ASK] do we really want to hardcode the nano X mac address ?

Comment (smartine-ledger) [reply bletransportsample/src/main/java/com/ledger/live/bletransportsample/screen/scan/ScanViewModel.kt]: [ASK] what don't we directly call `connectToDevice("DE:F1:55:51:C2:0C")` and so remove this method?

File: bletransportsample/src/main/java/com/ledger/live/bletransportsample/screen/scan/ScanViewModel.kt
Comment (smartine-ledger) [inline bletransportsample/src/main/java/com/ledger/live/bletransportsample/screen/scan/ScanViewModel.kt]: [ASK] Do we need two methods : `bleManager.disconnect()` and `bleManager.disconnect {...}`

File: transport-ble/src/main/java/com/ledger/live/ble/callback/BleManagerConnectionCallback.kt
Comment (smartine-ledger) [inline transport-ble/src/main/java/com/ledger/live/ble/callback/BleManagerConnectionCallback.kt:6]: [SHOULD] It's quiet confusing because BLE is for : Bluetooth Low Energy.

File: transport-ble/src/main/java/com/ledger/live/ble/model/BleDeviceService.kt
Comment (smartine-ledger) [inline transport-ble/src/main/java/com/ledger/live/ble/model/BleDeviceService.kt:12]: [ASK] Why `Builder` is a nested class of `BleDeviceService`?

File: transport-ble/src/main/java/com/ledger/live/ble/model/FrameCommand.kt
Comment (smartine-ledger) [inline transport-ble/src/main/java/com/ledger/live/ble/model/FrameCommand.kt:69]: [SHOULD] Not mix data class and utility class. You may have extension function instead

File: transport-ble/src/main/java/com/ledger/live/ble/BleManager.kt
Comment (smartine-ledger) [inline transport-ble/src/main/java/com/ledger/live/ble/BleManager.kt]: [SHOULD] create a custom scope with a context (and even a SupervisorJob if you need to use the same scope with several coroutines. (It's up to you) : `val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())`

File: transport-ble/src/main/java/com/ledger/live/ble/callback/BleManagerConnectionCallback.kt
Comment (smartine-ledger) [reply transport-ble/src/main/java/com/ledger/live/ble/callback/BleManagerConnectionCallback.kt:6]: [ERRATUM] After discussing we are in BLE context 👍

File: transport-ble/src/main/java/com/ledger/live/ble/callback/BleManagerDisconnectionCallback.kt
Comment (smartine-ledger) [inline transport-ble/src/main/java/com/ledger/live/ble/callback/BleManagerDisconnectionCallback.kt:3]: [SUGGESTiON] `fun interface BleManagerDisconnectionCallback{}` to allows this kind of call : 
`val result = BleManagerDisconnectionCallback{<execute action>}`

File: transport-ble/src/main/java/com/ledger/live/ble/callback/BleManagerSendCallback.kt
Comment (smartine-ledger) [inline transport-ble/src/main/java/com/ledger/live/ble/callback/BleManagerSendCallback.kt:3]: [SUGGESTiON] IMHO data class is a container of data. Unfortunately you have some logics here

File: transport-ble/src/main/java/com/ledger/live/ble/model/BleCommand.kt
Comment (smartine-ledger) [inline transport-ble/src/main/java/com/ledger/live/ble/model/BleCommand.kt:7]: hum... IMHO Single Responsibility Violation. What is the purpose of this class?

File: transport-ble/src/main/java/com/ledger/live/ble/service/BleService.kt
Comment (smartine-ledger) [inline transport-ble/src/main/java/com/ledger/live/ble/service/BleService.kt]: [COMMENT] As discussed you're working on this part. Do not hesitate if you want my opinion 🙂

File: transport-ble/src/main/java/com/ledger/live/ble/BleManager.kt
Comment (smartine-ledger) [inline transport-ble/src/main/java/com/ledger/live/ble/BleManager.kt]: [SHOULD] avoid using `runBlocking`because it  blocks the current thread

File: bletransportsample/src/main/java/com/ledger/live/bletransportsample/screen/scan/ScanViewModel.kt
Comment (mbertin-ledger) [reply bletransportsample/src/main/java/com/ledger/live/bletransportsample/screen/scan/ScanViewModel.kt]: It was a test to check that 2 consecutive disconnect was not crashing ^^'

**Comment by juan-cortes** [comment]
We really should merge this because github doesn't list branches other than the main one, please 🙏🏼 
cc @mbertin-ledger

**Comment by chandrucrypto** [comment]
@juan-cortes  @mrobert-ledger Lets please merge the PR and it would really be helpful to unblock the devs like me :-) I guarantee that I will also start contributing and refactor the code to make this repository awesome. For me all to do, this needs to be merged to `main` 😄

**Comment by mrobert-ledger** [comment]
> @juan-cortes @mrobert-ledger Lets please merge the PR and it would really be helpful to unblock the devs like me :-) I guarantee that I will also start contributing and refactor the code to make this repository awesome. For me all to do, this needs to be merged to `main` 😄

@chandrucrypto  Did you want to mention @mbertin-ledger instead of me ?

**Comment by mrobert-ledger** [comment]
> We really should merge this because github doesn't list branches other than the main one, please 🙏🏼 cc @mbertin-ledger

@juan-cortes GitHub do list all branches, even if you're not login, since it's a public repo. You can dl the branch feature/ble_library if needed

**Comment by juan-cortes** [comment]
> > We really should merge this because github doesn't list branches other than the main one, please 🙏🏼 cc @mbertin-ledger
> 
> @juan-cortes GitHub do list all branches, even if you're not login, since it's a public repo. You can dl the branch feature/ble_library if needed

What I meant by "doesn't list" is it doesn't index them in the search, I know I can download the repository to search through it but I can't easily search for things in this branch on github without it being merged first. Not a big deal, but would be nice.

**Comment by chandrucrypto** [comment]
@juan-cortes  @mbertin-ledger  @mrobert-ledger  The implementation  is good and the wallet connection works, but I still feel that we don't need compose. May be simple koin injection, view models, live datas, coroutines with Clean Architecture in place is enough at this point. I can modify it if you guys are fine with it.

**Comment by mbertin-ledger** [comment]
Hi @chandrucrypto , 
Thanks for showing interest on our library. 
As you can see we are at the first drafts of it and it's not totally ready. We put effort on the library rather than the sample that is not clean at all. So I understand you suggestion but keep in mind that the sample is pretty useless and as it's named, just show some example for using the library.  So, for the moment we will not implement livedatas, DI and make the sample right regarding clean architecture, as it's time consuming and the lib is not ready. 
If you look inside the library, you will be able to notice that we are using coroutine and most of the code is unit tested, and we are trying to respect SOLID principles.

Feel free to make suggestion once this branch will be merged, but for now as it is still in progress, I could be useless as we are still allowing us to make breaking change with the current implementation.

**Pull request merged**
